docs/design/sql/20-sql-partition-pruning.md

```diff
@@ -0,0 +1,309 @@
+  of Hazelcast. If other features are impacted, especially UX, list this impact as a reason not to do the change. If possible, also investigate and suggest mitigating actions that would reduce the impact. You can for example consider additional validation testing, additional documentation or doc changes, new user research, etc.
+- What parts of the design do you expect to resolve through the design process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this document that could be addressed in the future independently of the solution that comes out of this change?
+- How does this functionality impact/augment our Viridian Cloud offering? 
+- What changes, if any, are needed in Viridian to explose this functionality? 
+
+Use the ⚠️ or ❓icon to indicate an outstanding issue or question, and use the ✅ or ℹ️ icon to indicate a resolved issue or question.
+
+### User Interaction
+#### API design and/or Prototypes
+
+Listing of associated prototypes (latest versions) and any API design samples. How do we teach this?
+
+Explain the proposal as if it was already included in the project and you were teaching it to an end-user, or a Hazelcast team member in a different project area.
+
+Consider the following writing prompts:
+- Which new concepts have been introduced to end-users? Can you provide examples for each?
+- How would end-users change their apps or thinking to use the change?
+- Are there new error messages introduced? Can you provide examples?
+- Are there new deprecation warnings? Can you provide examples?
+- How are clusters affected that were created before this change? Are there migrations to consider?
+- How can this be leveraged in Viridian Cloud? 
+
+#### Client Related Changes
+Please identify if any client code change is required. If so, please provide a list of client code changes.
+The changes may include API changes, serialization changes or other client related code changes.
+
+Please notify the APIs team if any change is documented in this section. 
+The changes may need to be handled for non-Java clients as well.
+
+### Technical Design
+
+#### Assumptions
+
+##### Equi-partitions
+
+For IMap we support equality-based partitions which is inline with IMap being a hash table.
+Range-partitioning (eg. for date ranges) will be not supported.
+
+#### Partitioning metadata available to SQL optimizer
+
+Calcite SQL optimizer need to know how IMap is partitioned.
+
+List of columns comprising `AttributePartitioningStrategy` of the IMap will be available in `PartitionedMapTable`.
+(note that in this case entire `__key` also determines to which partition the entry belongs,
+but it must be evaluated using `AttributePartitioningStrategy` not general hash strategy).
+
+TODO: we want to benefit from the pruning also for `__key` in case `AttributePartitioningStrategy`.
```

Comment:
AFAIK `__key`, even if its available for querying etc in this context (where __key.attr is defined in the mapping that is), it would be of an OBJECT type, and we don't support OBJECT to ROW comparisons at all iirc.